### PR TITLE
Update CLI to update mapping only for spot/twap

### PR DESCRIPTION
**Pull request by silviutroscot**

### CODE REVIEW

File: off_chain/scope-cli/src/main.rs [hunk 1]
```diff
@@ -98,6 +99,10 @@ enum Actions {
         /// Where is stored the mapping to upload
         #[clap(long, env, parse(from_os_str))]
         mapping: PathBuf,
+
+        /// Mode to use for price upload
+        #[clap(long, env, default_value = "all")]
+        mode: PriceMode,
```
Comment (oeble) [inline off_chain/scope-cli/src/main.rs:105]: I don't really like "mode"; it is not explicit IMO.

File: Makefile [hunk 2]
```diff
@@ -180,9 +180,16 @@ init: $(SCOPE_CLI)
 
 update-mapping: $(SCOPE_CLI)
 >@ if [ $(CLUSTER) = "mainnet" ]; then\
-      RUST_BACKTRACE=1 RUST_LOG="scope_client=info" cargo run -p scope-cli -- --cluster $(URL) --multisig $(SCOPE_MULTISIG_AUTH) --program-id $(SCOPE_PROGRAM_ID) --price-feed $(FEED_NAME) upload --mapping ./configs/$(CLUSTER)/$(FEED_NAME).json;\
+      RUST_BACKTRACE=1 RUST_LOG="scope_client=info" cargo run -p scope-cli -- --cluster $(URL) --multisig $(SCOPE_MULTISIG_AUTH) --program-id $(SCOPE_PROGRAM_ID) --price-feed $(FEED_NAME) upload --mapping ./configs/$(CLUSTER)/$(FEED_NAME).json --mode $(MODE);\
   else\
-      RUST_BACKTRACE=1 RUST_LOG="scope_client=trace,scope=trace" cargo run -p scope-cli -- --cluster $(URL) --keypair $(OWNER_KEYPAIR) --program-id $(SCOPE_PROGRAM_ID) --price-feed $(FEED_NAME) upload --mapping ./configs/$(CLUSTER)/$(FEED_NAME).json;\
+      RUST_BACKTRACE=1 RUST_LOG="scope_client=trace,scope=trace" cargo run -p scope-cli -- --cluster $(URL) --keypair $(OWNER_KEYPAIR) --program-id $(SCOPE_PROGRAM_ID) --price-feed $(FEED_NAME) upload --mapping ./configs/$(CLUSTER)/$(FEED_NAME).json --mode $(MODE);\
+  fi
```
Comment (oeble) [inline Makefile:186]: You should not pass `--mode` if the variable to pass has the same name as this parameter have the `env` attribute anyway. By doing it like this, we can't call `make update-mapping` anymore without setting the env variable...

File: off_chain/scope-cli/src/scope_client.rs [hunk 3]
```diff
@@ -176,16 +177,34 @@ where
...
+                .filter(|(_, entry)| entry.get_label().contains("TWAP"))
+                .map(|(id, _)| *id)
+                .collect(),
+        };
+
         // For all "token" local and remote
-        for (&token_idx, local_entry) in &self.tokens {
+        // for (&token_idx, local_entry) in &tokens {
+        for &token_idx in &ids {
+            let local_entry = self.tokens.get(&token_idx).unwrap();
```
Comment (oeble) [inline off_chain/scope-cli/src/scope_client.rs:207]: Could you rewrite with the old iterator style but add a filter instead?

File: off_chain/scope-cli/src/scope_client.rs [hunk 4]
```diff
@@ -176,16 +177,34 @@ where
...
         let onchain_twap_enabled = program_mapping.twap_enabled;
         let onchain_twap_source = program_mapping.twap_source;
         let token_metadatas = self.get_token_metadatas().await?;
 
+        let ids: Vec<u16> = match mode {
+            PriceMode::All => self.tokens.keys().copied().collect(),
+            PriceMode::Spot => self
+                .tokens
+                .iter()
+                .filter(|(_, entry)| !entry.get_label().contains("TWAP"))
```
Comment (oeble) [inline off_chain/scope-cli/src/scope_client.rs:193]: This is an unreliable heuristic IMO. You should rather filter on the price type being `ScopeTwap`

File: off_chain/scope-cli/src/utils.rs [hunk 5]
```diff
@@ -30,3 +30,24 @@ pub async fn get_clock(rpc: &impl AsyncClient) -> Result<Clock> {
...
+
+    fn from_str(s: &str) -> std::prelude::v1::Result<Self, Self::Err> {
+        match s {
+            "all" => Ok(PriceMode::All),
+            "spot" => Ok(PriceMode::Spot),
+            "twap" => Ok(PriceMode::Twap),
+            _ => Err(anyhow::anyhow!("Invalid price mode")),
+        }
+    }
+}
```
Comment (oeble) [inline off_chain/scope-cli/src/utils.rs:53]: Please derive the [`clap::ValueEnum`](https://docs.rs/clap/latest/clap/trait.ValueEnum.html) instead. Also, add doc comments to each enum variant so the inline help is generated correctly for clap.

**Review by oeble — CHANGES_REQUESTED** [review]

File: Makefile [hunk 6]
```diff
@@ -180,9 +180,16 @@ init: $(SCOPE_CLI)
...
+
+update-mapping-spot: $(SCOPE_CLI)
+>@ if [ $(CLUSTER) = "mainnet" ]; then\
+      RUST_BACKTRACE=1 RUST_LOG="scope_client=info" cargo run -p scope-cli -- --cluster $(URL) --multisig $(SCOPE_MULTISIG_AUTH) --program-id $(SCOPE_PROGRAM_ID) --price-feed $(FEED_NAME) upload --mapping ./configs/$(CLUSTER)/$(FEED_NAME).json --mode twap;\
+  else\
+      RUST_BACKTRACE=1 RUST_LOG="scope_client=trace,scope=trace" cargo run -p scope-cli -- --cluster $(URL) --keypair $(OWNER_KEYPAIR) --program-id $(SCOPE_PROGRAM_ID) --price-feed $(FEED_NAME) upload --mapping ./configs/$(CLUSTER)/$(FEED_NAME).json --mode twap;\
```
[the remark this answers is not in the document]
Comment (silviutroscot) [reply Makefile:192]: removed as we can pass it to the generic `update-mapping`

**Review by oeble — APPROVED** [review]

**Pull request merged**